File: src/main/java/org/capturecoop/cccolorutils/gui/parts/CCAlphaBar.java
package org.capturecoop.cccolorutils.gui.parts;

import org.capturecoop.cccolorutils.CCColorUtils;
import org.capturecoop.ccutils.math.CCVector2Float;

import javax.imageio.ImageIO;
import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.MouseMotionAdapter;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.util.ArrayList;

public class CCAlphaBar extends JPanel {
    private Color backgroundColor;
    private float alpha;

    private final CCColorUtils.DIRECTION direction;

    private final static int MARGIN = 10;
    private final static int SEL_MARGIN = 4;
    private final static int SEL_MARGIN_OFF = 2;

    private static BufferedImage gridImage;

    private boolean isDragging = false;

    private BufferedImage buffer;
    private boolean dirty = true;
    private final ArrayList<ChangeListener> changeListeners = new ArrayList<>();

    public CCAlphaBar(Color color, CCColorUtils.DIRECTION direction, boolean alwaysGrab) {
        backgroundColor = color;
        this.direction = direction;
        alpha = color.getAlpha() / 255F;

        if(gridImage == null) {
            try {
                gridImage = ImageIO.read(CCAlphaBar.class.getResource("/org/capturecoop/cccolorutils/transparent_grid_4x4.png"));
            } catch (IOException ioException) {
                //Do nothing, that image is not essential.
            }
        }

        addMouseListener(new MouseAdapter() {
            @Override
            public void mousePressed(MouseEvent mouseEvent) {
                Rectangle rect = getSelectRect();
                if(rect == null)
                    return;

                if(rect.contains(mouseEvent.getPoint()) && !alwaysGrab)
                    isDragging = true;
            }

            @Override
            public void mouseReleased(MouseEvent mouseEvent) {
                isDragging = false;
                if(alwaysGrab)
                    execute(mouseEvent.getX(), mouseEvent.getY());
            }
        });

        addMouseMotionListener(new MouseMotionAdapter() {
            @Override
            public void mouseDragged(MouseEvent mouseEvent) {
                if(isDragging || alwaysGrab)
                    execute(mouseEvent.getX(), mouseEvent.getY());
            }
        });
    }

    private void execute(int x, int y) {
        int pos = y;
        int size = getHeight();
        if(direction == CCColorUtils.DIRECTION.HORIZONTAL) {
            pos = x;
            size = getWidth();
        }
        float percentage = (pos * 100F) / size;
        alpha = new CCVector2Float(percentage / 100F, 0).limitX(0, 1).getX();
        repaint();

        for(ChangeListener listener : changeListeners)
            listener.stateChanged(new ChangeEvent(this));
    }

    private int getSizeX() {
        return getWidth() - MARGIN;
    }

    private int getSizeY() {
        return getHeight() - MARGIN;
    }

    private Rectangle getSelectRect() {
        switch(direction) {
            case VERTICAL:
                int yPos = (int) (getSizeY() / (1 / alpha)) + MARGIN / 2;
                return new Rectangle(SEL_MARGIN_OFF, yPos - SEL_MARGIN, getWidth() - SEL_MARGIN_OFF * 2, SEL_MARGIN * 2);
            case HORIZONTAL:
                int xPos = (int) (getSizeX() / (1 / alpha)) + MARGIN / 2;
                return new Rectangle(xPos - SEL_MARGIN, SEL_MARGIN_OFF, SEL_MARGIN * 2, getHeight() - SEL_MARGIN_OFF * 2);
        }
        return null;
    }

    @Override
    public void paint(Graphics g) {
        if(buffer == null || !(buffer.getWidth() == getWidth() && buffer.getHeight() == getHeight())) {
            buffer = new BufferedImage(getWidth(), getHeight(), BufferedImage.TYPE_INT_ARGB);
        }

        int sizeX = getSizeX();
        int sizeY = getSizeY();
        g.setColor(getBackground());
        g.fillRect(0, 0, getWidth(), getHeight());
        if (dirty) {
            Graphics bufferGraphics = buffer.getGraphics();
            if (gridImage != null) {
                if (direction == CCColorUtils.DIRECTION.HORIZONTAL) {
                    int amount = sizeX / sizeY;
                    for (int i = 0; i < amount; i++) {
                        bufferGraphics.drawImage(gridImage, MARGIN / 2 + i * sizeY, MARGIN / 2, sizeY, sizeY, this);
                    }
                } else {
                    int amount = sizeY / sizeX;
                    for (int i = 0; i < amount; i++) {
                        bufferGraphics.drawImage(gridImage, MARGIN / 2, MARGIN / 2 + i * sizeX, sizeX, sizeX, this);
                    }
                }
            }
            bufferGraphics.drawImage(CCColorUtils.createAlphaBar(backgroundColor, sizeX, sizeY, direction), MARGIN / 2, MARGIN / 2, sizeX, sizeY, this);
            bufferGraphics.dispose();
            dirty = false;
        }
        g.drawImage(buffer, 0, 0, this);
        g.setColor(Color.BLACK);
        g.drawRect(MARGIN / 2 - 1, MARGIN / 2 - 1, sizeX + 1, sizeY + 1);
        g.setColor(Color.GRAY);
        Rectangle rect = getSelectRect();
        g.fillRect(rect.x, rect.y, rect.width, rect.height);
    }

    public void addChangeListener(ChangeListener listener) {
        //These only fire on changes made by the user
        changeListeners.add(listener);
    }

    public void setBackgroundColor(Color color) {
        backgroundColor = color;
        dirty = true;
        repaint();
    }

    public void setAlpha(int alpha) {
        this.alpha = alpha / 255F;
        dirty = true;
        repaint();
    }

    public int getAlpha() {
        return (int) (alpha * 255F);
    }
}

File: src/main/kotlin/org/capturecoop/cccolorutils/chooser/gui/CCAlphaBar.kt
package org.capturecoop.cccolorutils.chooser.gui

import org.capturecoop.cccolorutils.CCColorUtils.DIRECTION
import org.capturecoop.cccolorutils.CCColorUtils.createAlphaBar
import org.capturecoop.cccolorutils.fill
import org.capturecoop.ccutils.math.CCVector2Float
import java.awt.Color
import java.awt.Graphics
import java.awt.Rectangle
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.event.MouseMotionAdapter
import java.awt.image.BufferedImage
import javax.imageio.ImageIO
import javax.swing.JPanel
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener

class CCAlphaBar(private var backgroundColor: Color, private val direction: DIRECTION, alwaysGrab: Boolean) : JPanel() {
    private var alpha: Float
    private var isDragging = false
    private var buffer: BufferedImage? = null
    private var dirty = true
    private val changeListeners = ArrayList<ChangeListener>()

    init {
        alpha = backgroundColor.alpha / 255f
        gridImage = ImageIO.read(CCAlphaBar::class.java.getResource("/org/capturecoop/cccolorutils/transparent_grid_4x4.png"))

        addMouseListener(object : MouseAdapter() {
            override fun mousePressed(mouseEvent: MouseEvent) {
                if (selectRect.contains(mouseEvent.point) && !alwaysGrab) isDragging = true
            }

            override fun mouseReleased(mouseEvent: MouseEvent) {
                isDragging = false
                if (alwaysGrab) execute(mouseEvent.x, mouseEvent.y)
            }
        })
        addMouseMotionListener(object : MouseMotionAdapter() {
            override fun mouseDragged(mouseEvent: MouseEvent) {
                if (isDragging || alwaysGrab) execute(mouseEvent.x, mouseEvent.y)
            }
        })
    }

    private fun execute(x: Int, y: Int) {
        var pos = y
        var size = height
        if (direction === DIRECTION.HORIZONTAL) {
            pos = x
            size = width
        }
        val percentage = pos * 100f / size
        alpha = CCVector2Float(percentage / 100f, 0.0f).limitX(0f, 1f).x
        repaint()
        changeListeners.forEach { it.stateChanged(ChangeEvent(this)) }
    }

    private val sizeX: Int
        get() = width - MARGIN
    private val sizeY: Int
        get() = height - MARGIN
    private val selectRect: Rectangle
        get() = when (direction) {
            DIRECTION.VERTICAL -> {
                val yPos = (sizeY / (1 / alpha)).toInt() + MARGIN / 2
                Rectangle(SEL_MARGIN_OFF, yPos - SEL_MARGIN, width - SEL_MARGIN_OFF * 2, SEL_MARGIN * 2)
            }
            DIRECTION.HORIZONTAL -> {
                val xPos = (sizeX / (1 / alpha)).toInt() + MARGIN / 2
                Rectangle(xPos - SEL_MARGIN, SEL_MARGIN_OFF, SEL_MARGIN * 2, height - SEL_MARGIN_OFF * 2)
            }
        }


    override fun paint(g: Graphics) {
        if (buffer == null || !(buffer!!.width == width && buffer!!.height == height)) {
            buffer = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)
        }
        val sizeX = sizeX
        val sizeY = sizeY
        g.color = background
        g.fillRect(0, 0, width, height)
        if (dirty) {
            val bufferGraphics = buffer!!.graphics
            if (direction === DIRECTION.HORIZONTAL) {
                for (i in 0 until sizeX / sizeY)
                    bufferGraphics.drawImage(gridImage, MARGIN / 2 + i * sizeY, MARGIN / 2, sizeY, sizeY, this)
            } else {
                for (i in 0 until sizeY / sizeX)
                    bufferGraphics.drawImage(gridImage, MARGIN / 2, MARGIN / 2 + i * sizeX, sizeX, sizeX, this)
            }
            bufferGraphics.drawImage(createAlphaBar(backgroundColor, sizeX, sizeY, direction), MARGIN / 2, MARGIN / 2, sizeX, sizeY, this)
            bufferGraphics.dispose()
            dirty = false
        }
        g.drawImage(buffer, 0, 0, this)
        g.color = Color.BLACK
        g.drawRect(MARGIN / 2 - 1, MARGIN / 2 - 1, sizeX + 1, sizeY + 1)
        g.color = Color.GRAY
        selectRect.fill(g)
    }

    fun addChangeListener(listener: ChangeListener) {
        //These only fire on changes made by the user
        changeListeners.add(listener)
    }

    fun setBackgroundColor(color: Color) {
        backgroundColor = color
        dirty = true
        repaint()
    }

    fun setAlpha(alpha: Int) {
        this.alpha = alpha / 255f
        dirty = true
        repaint()
    }

    fun getAlpha(): Int {
        return (alpha * 255f).toInt()
    }

    companion object {
        private const val MARGIN = 10
        private const val SEL_MARGIN = 4
        private const val SEL_MARGIN_OFF = 2
        private lateinit var gridImage: BufferedImage
    }
}